ikarus/finiteelements/fehelper.hh

```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
         localX[i][j] = x[localView.index(localView.tree().child(j).localIndex(i))[0]];
   return localX;
 }
+
+/**
+ * \brief A function which loops over all the nodes of a tree and performs different actions
+ * for a power node and a leaf node depending on the corresponding functor passed.
+ *
+ * \details This function is inspired from the function Dune::TypeTree::Detail::forEachNode
+ * available in dune/typetree/traversal.hh
+ *
+ * \tparam T Type of tree.
+ * \tparam ChildIndex Type of the index of the child.
+ * \tparam PowerFunc Type of the functor called for a power node.
+ * \tparam LeafFunc Type of the functor called for a leaf node.
+ * \param tree Tree whose nodes are to be looped over.
+ * \param childIndex The index of the child.
+ * \param powerFunc A functor to be called for a power node.
+ * \param leafFunc A functor to be called for a leaf node.
+ */
+template <class T, class ChildIndex, class PowerFunc, class LeafFunc>
+void forEachLeafOrPowerNode(T&& tree, ChildIndex&& childIndex, PowerFunc&& powerFunc, LeafFunc&& leafFunc) {
+  using Tree = std::decay_t<T>;
+  if constexpr (Tree::isLeaf) {
+    leafFunc(tree);
+  } else if constexpr (Tree::isPower) {
+    if constexpr (Tree::template Child<Dune::Indices::_0>::Type::isComposite) {
```

Comment:
This is a strange statement it should be the other way around. Check if the child of the powernode is leaf. If not call forEachLeafOrPowerNode again with the child.

Reply:
If I check if the child of a power node is a leaf and call the `leafFunc`, then it will no longer be in `FlatInterleaved` format right? Because then each leaf node of a power node will be looped over bringing in `x0,x1,...,xn,y0,y1,...,yn` format? In order to invoke `powerFunc` at a power node directly and have `FlatInterleaved` format, I chose the other way around of checking if it is not a leaf. May be we can discuss this together.

Reply:
Not you should check if the child of the power node is leaf and then pass the power node.

Reply:
Sinece we only want to call the `powerNodeOp` for power nodes, which has leaf children. So only for `power(lagrange,d)` and  `power(nedelec,d)`  but not for `power(composite(nedelec, power(lagrange,d2)),d)`

Reply:
Then if you pass the power node int o`powerNodeOp` you only implement the `if` branch. We can also rename the function name to `forEachLeafOrPowerLeafNode`